src/neighborsearch/filtered.jl

```diff
@@ -0,0 +1,193 @@
+end
+
+maxneighbors(method::FilteredSearch) = method.nmax
+
+function search!(neighbors, pₒ::Point, method::FilteredSearch; mask=nothing)
+  meth = method.method
+  obj  = meth.domain
+  nmax = method.nmax
+
+  # initial search, using mask if necessary
+  inds = search(pₒ, meth, mask=mask)
+
+  # `inds` is not guranteed to be sorted by distance. also, a different
+  # metric might be preferred to give priority to the FilteredSearch.
+  # unless `metric = nothing`, the indices are sorted here using metric
+  if !isnothing(method.metric)
+    dists  = [evaluate(method.metric, coordinates(pₒ), coordinates(centroid(obj, ind))) for ind in inds]
+    sorted = sortperm(dists)
+    inds   = inds[sorted]
+  end
+
+  # initialize category and sectors constraints if necessary. `Dict` containers
+  # are here created with the information necessary for further filtering.
+  categs  = initcategories(obj, inds, method.maxpercategory)
+  sectors = initsectors(meth, method.maxpersector)
+
+  # loop each neighbor candidate
+  nneigh = 0
+  for i in inds
+    # check category. if exceeded the category, the neighbor is ignored;
+    # otherwise categs[:count][col][cat[col]] is incremented in the end after
+    # nmax and maxpersector checks
+    if categs[:use]
+      cat, pass = Dict(), false
+      tab = Tables.columns(values(obj))
+      for col in keys(categs[:max])
+        cat[col] = Tables.getcolumn(tab, col)[i]
+        categs[:count][col][cat[col]] >= categs[:max][col] && (pass = true)
+      end
+      pass && continue
+    end
+
+    # check sectors; if there are no space for neighbors in the indsector, it's
+    # ignored; otherwise sectors[:count][indsector] is incremented in the end
+    # after nmax checks
+    if sectors[:use]
+      centered  = sectors[:rotmat]' * (centroid(obj,i) - pₒ)
+      indsector = getsector(centered)
+      sectors[:count][indsector] >= sectors[:max] && continue
+    end
+
+    # add neighbor
+    nneigh += 1
+    if nmax == Inf
+      push!(neighbors, i)
+    else
+      neighbors[nneigh] = i
+    end
+
+    nneigh == nmax && break
+
+    # add counters
+    sectors[:use] && (sectors[:count][indsector] += 1)
+    if categs[:use]
+      for col in keys(categs[:max])
+        categs[:count][col][cat[col]] += 1
+      end
+    end
+  end
+
+  # slice neighbors if nmax was not reached
+  length(neighbors) > nneigh && (neighbors = neighbors[1:nneigh])
+
+  nneigh
+end
+
+# initialize categories constraints if necessary. e.g. if `maxpercategory =
+# (holeid = 2,)`, dict[:count][:holeid] will have all the possible values of
+# holeid, initiliazed to zero for further counting.
+function initcategories(obj, inds, catgs)
+  catgs == nothing && return Dict(:use => false)
+  tab   = Tables.columns(values(obj))
+  catvals = Dict(k => unique(Tables.getcolumn(tab, k)[inds]) for k in keys(catgs))
+  counter = Dict(k => Dict(zip(v, zeros(Int,length(v)))) for (k, v) in catvals)
+  Dict(:use => true, :max => catgs, :count => counter)
+end
```

Comment:
We need to refactor this. Dictionaries don't seem to be the correct data structure to count categories. Also, we have too much nesting in this dictionary, which is an indication that the design should be reconsidered.

Let's think about the original problem. We start with a set of candidate `inds` returned by the inner method. Now we want to filter these indices such that a maximum number of indices is retained for each specified categorical property. So we need to define a filtering operation. That is clear.

The way we write this filter is the key here. Let's introduce a counter for each specified categorical property:

```julia
countpercat = Dict(keys(maxpercategory) .=> 0)
```

For each candidate index, we check if it will be retained by checking if all the counts did not surpass the maximum counts allowed:

```julia
filtered = Int[]
for ind in inds
  retain = true
  for (k, v) in maxpercategory
    if countpercat[k] >= v
      retain = false
    end
  end
  if retain
    push!(filtered, ind)
    # update countpercat by inspecting the line `ind` only
  end
end
```

I think this pseudo-code should be enough. Can you double check?